google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,936 @@
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+
+	cluster, err := config.clientDataproc.Projects.Regions.Clusters.Get(
+		project, region, clusterName).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataproc Cluster %q", clusterName))
+	}
+
+	d.Set("name", cluster.ClusterName)
+	d.Set("region", region)
+	d.Set("labels", cluster.Labels)
+
+	cfg, err := flattenClusterConfig(d, cluster.Config)
+	if err != nil {
+		return err
+	}
+
+	d.Set("cluster_config", cfg)
+	return nil
+}
+
+func flattenClusterConfig(d *schema.ResourceData, cfg *dataproc.ClusterConfig) ([]map[string]interface{}, error) {
+
+	data := getOrCreateNewMapSchema(d, "cluster_config")
+
+	data["bucket"] = cfg.ConfigBucket
+	data["gce_cluster_config"] = flattenGceClusterConfig(data, cfg.GceClusterConfig)
+	data["software_config"] = flattenSoftwareConfig(data, cfg.SoftwareConfig)
+
+	data["master_config"] = flattenInstanceGroupConfig(data, "master_config", cfg.MasterConfig)
+	data["worker_config"] = flattenInstanceGroupConfig(data, "worker_config", cfg.WorkerConfig)
+	data["preemptible_worker_config"] = flattenPreemptibleInstanceGroupConfig(data, "preemptible_worker_config", cfg.SecondaryWorkerConfig)
+
+	if len(cfg.InitializationActions) > 0 {
+		val, err := flattenInitializationActions(cfg.InitializationActions)
+		if err != nil {
+			return nil, err
+		}
+		data["intialization_action"] = val
+	}
+	return []map[string]interface{}{data}, nil
+}
+
+func flattenSoftwareConfig(parent map[string]interface{}, sc *dataproc.SoftwareConfig) []map[string]interface{} {
+	data := getOrCreateNewMap(parent, "software_config")
+	data["image_version"] = sc.ImageVersion
+	data["properties"] = sc.Properties
+
+	return []map[string]interface{}{data}
+}
+
+func flattenInitializationActions(nia []*dataproc.NodeInitializationAction) ([]map[string]interface{}, error) {
+
+	actions := []map[string]interface{}{}
+	for _, v := range nia {
+		action := map[string]interface{}{
+			"script": v.ExecutableFile,
+		}
+		if len(v.ExecutionTimeout) > 0 {
+			tsec, err := extractInitTimeout(v.ExecutionTimeout)
+			if err != nil {
+				return nil, err
+			}
+			action["timeout_sec"] = tsec
+		}
+
+		actions = append(actions, action)
+	}
+	return actions, nil
+
+}
+
+func flattenGceClusterConfig(parent map[string]interface{}, gcc *dataproc.GceClusterConfig) []map[string]interface{} {
+
+	gceConfig := getOrCreateNewMap(parent, "gce_cluster_config")
+	gceConfig["zone"] = extractLastResourceFromUri(gcc.ZoneUri)
+	gceConfig["tags"] = gcc.Tags
+	gceConfig["service_account"] = gcc.ServiceAccount
+
+	if gcc.NetworkUri != "" {
+		gceConfig["network"] = extractLastResourceFromUri(gcc.NetworkUri)
+	}
+	if gcc.SubnetworkUri != "" {
+		gceConfig["subnetwork"] = extractLastResourceFromUri(gcc.SubnetworkUri)
+	}
+
+	if len(gcc.ServiceAccountScopes) > 0 {
+		sort.Strings(gcc.ServiceAccountScopes)
```

Comment:
Won't this lead to diffs if the user didn't originally specify them alphabetically?